File: ocaml/parser.ml
(* Parser implementation.
 * Currently supports operation precedence and grouping by associativity. *)

open Core

module M = Mark
module L = Lexer
module S = Symbol

exception ParserError of string

(* Defines operation associativity. *)
type assoc = Left | Right

(* 
 * atom : int | symbol | lparen * exp * lparen
 * exp : atom * op * exp 
 *)

let parse lexer =
  let op_info op = 
    match op with
    | L.Op.Pow -> (3, Right)
    | L.Op.Plus | L.Op.Minus -> (1, Left)
    | L.Op.Times | L.Op.Divide -> (2, Left)
    | _ -> (100, Left)
  in

  let rec parse_expr min_prec =
    let lhs = parse_atom () in
    let result = parse_expr_prec min_prec lhs in
    result
  and parse_expr_prec min_prec lhs = 
    let mtok = L.peek lexer in
    let tok = M.obj mtok in
    match tok with
    | L.Delim | L.Eof | L.IntVal _| L.BoolVal _ -> L.drop lexer; lhs
    | L.Operator op ->
      let prec, assoc = op_info op in
      if prec < min_prec then lhs
      else 
        (* Only consume the token if it will be used in the AST *)
        let () = L.drop lexer in
        (* For left-associativity, increase precedence to left-group items *)
        let next_min_prec = 
          match assoc with
          | Left -> prec + 1
          | Right -> prec
        in
        let rhs = parse_expr next_min_prec in
        let exp = Mark.create_from_range lhs rhs (Ast.Operator (op, lhs, rhs))
        in parse_expr_prec min_prec exp
    | _ -> lhs
  and parse_atom () =
    let mtok = L.pop lexer in
    let tok = Mark.obj mtok in
    match tok with
    | L.IntVal i -> Mark.with_mark (Ast.IntVal i) mtok
    | L.BoolVal b -> Mark.with_mark (Ast.BoolVal b) mtok
    | L.Symbol s -> Mark.with_mark (Ast.Var (S.create s)) mtok
    | L.LParen ->
      let e = parse_expr 0 in
      let mtok2 = L.pop lexer in
      let tok2 = Mark.obj mtok2 in
      (match tok2 with
       | L.RParen -> e
       | _ -> 
         Err.print (Lexer.file lexer) (Lexer.fname lexer) mtok2;
         raise (ParserError "Unbalanced parentheses"))
    | _ -> 
      Err.print (Lexer.file lexer) (Lexer.fname lexer) mtok;
      raise (ParserError "Illegal atom grammar")
  in
  parse_expr 0

let rec parse_stmt lexer = 
  let mtok = L.pop lexer in
  let tok = Mark.obj mtok in
  let stmt = 
    match tok with
    | L.Symbol s -> 
      let meq = L.pop lexer in
      let eq = Mark.obj meq in
      let assign = 
        match eq with
        | L.Eq -> Ast.Assign (S.create s, parse lexer)
        | _ -> 
          Err.print (Lexer.file lexer) (Lexer.fname lexer) meq;
          raise (ParserError "var assignment: should be followed by =")
      in
      Mark.with_mark assign mtok
    | L.Return -> Mark.with_mark (Ast.Return (parse lexer)) mtok
    (* This only exists for empty-line commands. *)
    | L.Delim -> parse_stmt lexer
    | L.BoolDecl ->
      let mvar = L.peek lexer in
      let var = Mark.obj mvar in
      let decl =
        match var with
        | L.Symbol s -> Ast.Declare(Type.Bool, S.create s)
        | _ -> 
          Err.print (Lexer.file lexer) (Lexer.fname lexer) mvar;
          raise (ParserError "var declaration: keyword bool should be followed by symbol")
      in
      Mark.with_mark decl mtok
    | L.IntDecl ->
      let mvar = L.peek lexer in
      let var = Mark.obj mvar in
      let decl =
        match var with
        | L.Symbol s -> Ast.Declare(Type.Int, S.create s)
        | _ -> 
          Err.print (Lexer.file lexer) (Lexer.fname lexer) mvar;
          raise (ParserError "var declaration: keyword int should be followed by symbol")
      in
      Mark.with_mark decl mtok
    | _ -> 
      Err.print (Lexer.file lexer) (Lexer.fname lexer) mtok;
      raise (ParserError "statement does not begin with assignment or return")
  in stmt

let parse_program lexer =
  let rec parse_program' prog = 
    let mtok = L.peek lexer in
    let tok = Mark.obj mtok in
    match tok with
    | L.Eof -> prog
    | _ -> let stmt = parse_stmt lexer in parse_program' (List.append prog [stmt])
  in
  parse_program' []

File: go/parser/parser.go
package parser

import (
	"errors"

	"github.com/eyluo/jelly/ast"
	"github.com/eyluo/jelly/lexer"
	"github.com/eyluo/jelly/symbol"
)

type assoc int

const (
	left assoc = iota
	right
)

func Parse(lxr *lexer.T) (*ast.Exp, error) {
	var parseExpr func(minPrec int) (*ast.Exp, error)
	var parseExprPrec func(minPrec int, lhs *ast.Exp) (*ast.Exp, error)
	var parseAtom func() (*ast.Exp, error)

	opInfo := func(op lexer.OpType) (int, assoc) {
		var prec int
		var assc assoc
		switch op {
		case lexer.Pow:
			prec, assc = 3, right
		case lexer.Plus:
			fallthrough
		case lexer.Minus:
			prec, assc = 1, left
		case lexer.Times:
			fallthrough
		case lexer.Divide:
			prec, assc = 2, left
		}
		return prec, assc
	}

	parseExpr = func(minPrec int) (*ast.Exp, error) {
		lhs, err := parseAtom()
		if err != nil {
			return nil, err
		}

		result, err := parseExprPrec(minPrec, lhs)
		if err != nil {
			return nil, err
		}

		return result, nil
	}

	parseExprPrec = func(minPrec int, lhs *ast.Exp) (*ast.Exp, error) {
		var result *ast.Exp
		tok := lxr.Peek()
		switch tok.Type {
		case lexer.Delim:
			fallthrough
		case lexer.Eof:
			fallthrough
		case lexer.IntVal:
			lxr.Drop()
			result = lhs
		case lexer.Operator:
			prec, assc := opInfo(tok.Optr)
			if prec < minPrec {
				return lhs, nil
			}

			lxr.Drop()

			var nextMinPrec int
			if assc == left {
				nextMinPrec = prec + 1
			} else {
				nextMinPrec = prec
			}

			rhs, err := parseExpr(nextMinPrec)
			if err != nil {
				return nil, err
			}

			temp := &ast.Exp{
				Type: ast.Operator,
				Optr: tok.Optr,
				E1:   lhs,
				E2:   rhs,
			}

			return parseExprPrec(minPrec, temp)
		default:
			result = lhs
		}

		return result, nil
	}

	parseAtom = func() (*ast.Exp, error) {
		tok := lxr.Pop()
		switch tok.Type {
		case lexer.IntVal:
			return &ast.Exp{
				Type: ast.IntVal,
				Val:  tok.Val,
			}, nil
		case lexer.Symbol:
			return &ast.Exp{
				Type: ast.Var,
				Sym:  symbol.New(tok.Sym),
			}, nil
		case lexer.LParen:
			exp, err := parseExpr(0)
			if err != nil {
				return nil, err
			}

			tok2 := lxr.Pop()
			if tok2.Type != lexer.RParen {
				return nil, errors.New("unbalanced parentheses")
			}

			return exp, nil
		default:
			return nil, errors.New("illegal atom")
		}
	}

	return parseExpr(0)
}

func ParseStmt(lxr *lexer.T) (*ast.Stmt, error) {
	tok := lxr.Pop()
	switch tok.Type {
	case lexer.Symbol:
		eq := lxr.Pop()
		switch eq.Type {
		case lexer.Eq:
			exp, err := Parse(lxr)
			if err != nil {
				return nil, err
			}

			return &ast.Stmt{
				Type: ast.Assign,
				Sym:  symbol.New(tok.Sym),
				Exp:  exp,
			}, nil
		default:
			return nil, errors.New("assignment should be followed by =")
		}
	case lexer.Return:
		exp, err := Parse(lxr)
		if err != nil {
			return nil, err
		}

		return &ast.Stmt{
			Type: ast.Return,
			Exp:  exp,
		}, nil
	case lexer.Delim:
		return ParseStmt(lxr)
	default:
		return nil, errors.New("statement does not begin with assignment or return")
	}
}

func ParseProgram(lxr *lexer.T) (ast.Program, error) {
	tok := lxr.Peek()
	switch tok.Type {
	case lexer.Eof:
		return make([]*ast.Stmt, 0), nil
	default:
		stmt, err := ParseStmt(lxr)
		if err != nil {
			return nil, err
		}

		result, err := ParseProgram(lxr)
		if err != nil {
			return nil, err
		}

		return append([]*ast.Stmt{stmt}, result...), nil
	}
}
